mmdet3d/models/detectors/imvotenet.py

```diff
@@ -1,51 +1,129 @@
+import numpy as np
+import torch
 from torch import nn as nn
+from torch.nn import functional as F
 
-from mmdet.models import DETECTORS, build_backbone, build_head, build_neck
+from mmdet3d.core import bbox3d2result, merge_aug_bboxes_3d
+from mmdet.models import DETECTORS
+from .. import builder
 from .base import Base3DDetector
 
 
+class ImageMLPModule(nn.Module):
+
+    def __init__(self, input_dim=18, image_hidden_dim=256):
+        super().__init__()
+        self.img_feat_conv1 = nn.Conv1d(input_dim, image_hidden_dim, 1)
+        self.img_feat_conv2 = nn.Conv1d(image_hidden_dim, image_hidden_dim, 1)
+        self.img_feat_bn1 = nn.BatchNorm1d(image_hidden_dim)
+        self.img_feat_bn2 = nn.BatchNorm1d(image_hidden_dim)
+
+    def forward(self, img_features):
+        img_features = F.relu(
+            self.img_feat_bn1(self.img_feat_conv1(img_features)))
+        img_features = F.relu(
+            self.img_feat_bn2(self.img_feat_conv2(img_features)))
+
+        return img_features
+
+
+def sample_valid_seeds(mask, num_sampled_seed=1024):
+    device = mask.device
+    batch_size = mask.shape[0]
+    sample_inds = torch.zeros((batch_size, num_sampled_seed), device=device)
+    for bidx in range(batch_size):
+        # return index of non zero elements
+        valid_inds = torch.arange(len(mask[bidx, :]))
+        if len(valid_inds) < num_sampled_seed:
+            rand_inds = np.random.choice(
+                list(
+                    set(np.arange(num_sampled_seed)) -
+                    set(np.mod(valid_inds, num_sampled_seed))),
+                num_sampled_seed - len(valid_inds),
+                replace=False)
+            rand_inds = torch.new_tensor(rand_inds, device=device)
+            cur_sample_inds = torch.cat((valid_inds, rand_inds))
+        else:
+            cur_sample_inds = np.random.choice(
+                valid_inds, num_sampled_seed, replace=False)
+            cur_sample_inds = torch.new_tensor(cur_sample_inds, device=device)
+        sample_inds[bidx, :] = cur_sample_inds
+    return sample_inds.long()
+
+
 @DETECTORS.register_module()
 class ImVoteNet(Base3DDetector):
     r"""`ImVoteNet <https://arxiv.org/abs/2001.10692>`_ for 3D detection."""
 
     def __init__(self,
                  pts_backbone=None,
-                 pts_bbox_head=None,
+                 pts_bbox_head_joint=None,
```

Comment:
can we make pts_bbox_head a list or a dict of config dicts?

Reply:
I plan to change it to a dict which consists of the common parts and the parts that belong to each head